### PR TITLE
Adds has_downloads and last_download_removed fields to UserSyncStatus API response

### LINKED ISSUE #10378: Add UserSyncStatus API response fields for downloads and removals
Depends on: https://github.com/learningequality/kolibri/pull/10365

## Background
In order to complete Learner notifications for automatic content syncing, we need additional data points made available with work from #10365. This will allow us to wire up the [already implemented notifications](https://github.com/learningequality/kolibri/pull/10056/files#diff-155b7b2af6eb8f44179578e0b3670d02f3938df8170a9a2bcdaea7ff196dbcb7R71-R72) with this new data.

### The [`UserSyncStatus` API](https://github.com/learningequality/kolibri/blob/develop/kolibri/core/device/api.py#L284)
a) It should have a new response field `has_downloads` which annotates the existence of `ContentDownloadRequest` models for the user. For example:
```python
Exists(
  ContentDownloadRequest.objects.filter(
    source_id=OuterRef("user_id"), 
    source_model=FacilityUser.morango_model_name, 
    reason=ContentRequestReason.UserInitiated,
  )
)
```
b) It should have a new response field `last_download_removed` which annotates the date of the most recent `ContentRemovalRequest` model for the user with the `SyncInitiated` reason. For example:
```python
Subquery(
  ContentRemovalRequest.objects.filter(
    source_id=OuterRef("user_id"), 
    source_model=FacilityUser.morango_model_name, 
    reason=ContentRequestReason.SyncInitiated,
  )
  .annotate(last_removal=Max("requested_at"))
  .values("last_removal")
)
```

### The composable [`useUserSyncStatus`](https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/composables/useUserSyncStatus.js)
a) It should be updated to extract the newly added fields `has_downloads` and `last_download_removed`

### The learner Vue component [`StorageNotification.vue`](https://github.com/AllanOXDi/kolibri/blob/develop/kolibri/core/assets/src/views/StorageNotification.vue#L67-L72)
a) It should be refactored to use the `useUserSyncStatus` composable internally, instead of accepting a `showBanner` prop from `AppBarPage.vue`
b) It should update usage of `insufficientSpace` to match `deviceStatus === LearnerDeviceStatus.INSUFFICIENT_STORAGE`
c) It should update usage of `availableDownloads` to match the new API response field `has_downloads`
d) It should update usage of `resourcesRemoved` to match the new API response field `last_download_removed`
e) It should update usage of `hasDevicePermission` to use [`canManageContent`](https://github.com/learningequality/kolibri/blob/develop/kolibri/core/assets/src/state/modules/session.js#L38) from the user session
f) It should persist the user's dismissal of the notification, using the dates of the last sync and the last removed download

## Acceptance criteria
The notification should properly render the notification as designed in the Figma spec https://www.figma.com/file/MpCG7r5PULOCJsEvnTSti8/Android-Designs-2022?node-id=194-2922

## References
https://github.com/learningequality/kolibri/pull/10056
https://github.com/learningequality/kolibri/issues/10184

### PR DESCRIPTION
**Pull request by AllanOXDi**
## Summary

- Added a new response field called` has_download `and `last_download_removed.`
- Updated the useUserSyncStatus composable to extract the newly added fields` has_downloads `and `last_download_removed` from the UserSyncStatus API response.
- Refactored [StorageNotification.vue](https://github.com/AllanOXDi/kolibri/blob/develop/kolibri/core/assets/src/views/StorageNotification.vue#L67-L72) component to utilize the useUserSyncStatus composable internally, eliminating the need for the showBanner prop from AppBarPage.vue.
- Modified the usage of insufficientSpace to align with `deviceStatus === LearnerDeviceStatus.INSUFFICIENT_STORAGE.`
- Adapted the usage of availableDownloads to reflect the new `has_downloads `field in the API response.
- Adjusted the usage of `resourcesRemoved `to correspond to the new `last_download_removed `field in the API response.
- Updated the usage of `hasDevicePermission` to utilize` canManageContent` from the user session.
closes  #10378

## References
 #10378

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/device/api.py
Comment (bjester) [inline kolibri/core/device/api.py]: :eyes:

Comment (bjester) [reply kolibri/core/device/api.py]: As your test confirms, you've put forth some testing against this code. Although, I don't see `pdb` imported, leading me to suspect this code should raise an error if ever encountered. If it hasn't errored, then that must mean it isn't being invoked!

Comment (bjester) [reply kolibri/core/device/api.py]: This is actually a useful debugging practice during test development. If you want to sanity check an all-good, passing test result, you can throw an error, i.e. `raise new Exception('oops')`, which you expect to cause a clear failure.

Comment (AllanOXDi) [reply kolibri/core/device/api.py]: Looks like I just forgot to remove the `pdb` before pushing the code.

**Comment by AllanOXDi** [comment]
Hi Reviewers, 
I kindly request a few moments of your time to re-review this again.

**Comment by bjester** [comment]
Sure thing @AllanOXDi, I'll give it a re-review first thing tomorrow!

**Comment by bjester** [comment]
@AllanOXDi Seems at least one of the deliverables in the issue is still incomplete:
> f) It should persist the user's dismissal of the notification, using the dates of the last sync and the last removed download

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (bjester) [inline kolibri/core/assets/src/views/StorageNotification.vue]: I don't quite understand this change? Could you clarify what you're intending here. It seems this would mean that if there's insufficient storage, the notification isn't dismissable

Comment (AllanOXDi) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Yes! That was my thought.

Comment (bjester) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Did you discuss this with anyone? I feel that might be unexpected behavior

Comment (AllanOXDi) [reply kolibri/core/assets/src/views/StorageNotification.vue]: I did not discuss it with anyone.  I am not sure of what would be the expected behavior here. if that's the case.

Comment (bjester) [reply kolibri/core/assets/src/views/StorageNotification.vue]: > f) It should persist the user's dismissal of the notification, using the dates of the last sync and the last removed download

So I think this connects with this deliverable above that I mentioned.

There are two primary situations where we want this notification to show:
1) The device has insufficient storage
2) User's downloads were removed

I think the text of the notification should show in that priority: if both are true, (1) should show. Either way, the notification should be dismissable. We can track the user's dismissal in the browser's local storage (see this [Vue-use composable](https://vueuse.org/core/useLocalStorage/)). So when we query the API via `useUserSyncStatus`, we should compare what's stored in local storage with those details. Specifically, we can utilize the dates of the last sync and the last removed download to know whether the user has dismissed the notification or not.

Comment (bjester) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Remember to take a look at the figma for the different use cases. I overlooked some things related to the full-device scenarios, but I think the frontend logic should be the same regardless.
Screenshot from 2023-06-01 08-09-41

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (akolson) [inline kolibri/core/assets/src/views/StorageNotification.vue]: Not sure if this is an appropriate name to to use for this method; it seems a little misleading based on its logic. Maybe maintaining `toggleBanner` would suffice?

Comment (akolson) [reply kolibri/core/assets/src/views/StorageNotification.vue]: I also think the logic therein is still be flawed. If `showBanner` is true [which I presume is meant to display the banner], setting `bannerOpened` to `false` would dismiss it instead. I think it should be the other way round. 
if my assumption is correct, a simple `this.bannerOpened = this.showBanner` would be enough to display the banner when the condition to show a banner is met.

**Review by akolson — DISMISSED** [review]
Thanks @AllanOXDi! Code generally LGTM. However, I am not not too sure the issues raised in this [conversation](https://github.com/learningequality/kolibri/pull/10709#issuecomment-1572107650) have been fully resolved. I will be waiting on @bjester for a final review. Also, I left a few comments for you to look into.

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (bjester) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Part of the reason I mentioned the Vue-use composable for local storage was that it should also be reactive to changes. So if the component was using it, the computed prop `showBanner` would change to false, when setting `download_removed` and `last_synced` in local storage. That should make managing `bannerOpened` no longer needed

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/StorageNotification.vue:58]: We should continue to use this rather than the isSubsetOfUsersDevice below. This will avoid weird interactions after one user has migrated to a learner only device, while others have not.

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/StorageNotification.vue]: @bjester had suggested using this https://vueuse.org/core/useLocalStorage/ in order to make sure that these localStorage references are reactive, so that this computed prop will be recalculated.

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (rtibbles) [inline kolibri/core/assets/src/views/StorageNotification.vue]: Not suggesting a change here, just a point of interest for you. Another way of handling this would be to use the getter setter pattern for the showBanner computed prop. More info here (but no need to change this!): https://v2.vuejs.org/v2/guide/computed.html#Computed-Setter

File: kolibri/core/content/models.py
Comment (rtibbles) [inline kolibri/core/content/models.py:460]: This seems like an important change! @marcellamaki do you have a similar change in your PR?

File: kolibri/core/device/test/test_api.py
Comment (rtibbles) [inline kolibri/core/device/test/test_api.py]: My only other thought here is that it might be good to have test case where you expect this to be True and for last_download_removed not to be None.

**Review by rtibbles — COMMENTED** [review]
We should keep isLearnerOnlyImport. Made a couple of other comments and suggestions. Would like another eye from either @bjester or @akolson in case I have missed anything that needs resolution!

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (akolson) [inline kolibri/core/assets/src/views/StorageNotification.vue]: I think the `useLocalStorage` referred to here is https://vueuse.org/core/useLocalStorage/#uselocalstorage

Comment (akolson) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Also, please note the camel case 🙂

File: kolibri/core/device/test/test_api.py
Comment (akolson) [inline kolibri/core/device/test/test_api.py]: It looks like the assert suggested by Richard has not been added here

**Review by akolson — COMMENTED** [review]
Thanks @AllanOXDi. Just a few more comments(Particularly that coming from Richard) to address and we should be good I think. @bjester ?

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (bjester) [inline kolibri/core/assets/src/views/StorageNotification.vue]: Looking at the `vueuse` docs, it appears the return value from `useLocalStorage` is already a `Ref`. The [docs](https://v9-13-0.vueuse.org/core/useLocalStorage/) show it returns `RemovableRef` under 'Type Declarations':
```ts
export declare function useLocalStorage(
  key: string,
  initialValue: MaybeRefOrGetter<string>,
  options?: UseStorageOptions<string>
): RemovableRef<string>
```

The ending part `: RemovableRef<string>` says `useLocalStorage` returns the `RemovableRef` type, which can hold values of type `string` (the part inside `<` `>`)

Admittedly, it was hard to track whether `RemovableRef` is a sub-class/type of `Ref`, but it appears so:
https://github.com/vueuse/vueuse/blob/bddfbae423426ea70a2cf961350562c2b7758cd6/packages/shared/utils/types.ts#L13-L19

File: kolibri/core/content/models.py
Comment (bjester) [reply kolibri/core/content/models.py:460]: Yeah I mentioned this on Slack

**Review by bjester — COMMENTED** [review]
Code read through looks good. Just one comment about the usage of `ref(...)`. I will do some testing on it. Great perseverance on this @AllanOXDi !

File: kolibri/core/assets/src/views/StorageNotification.vue
Comment (akolson) [reply kolibri/core/assets/src/views/StorageNotification.vue]: Based on @AllanOXDi's request, I will be stepping in to resolve any pending issues on this pr while he is away.

**Comment by akolson** [comment]
Changes as suggested by @bjester  have been incorporated and with a final look through, we should be good to merge this pr.

File: kolibri/core/device/test/test_api.py
Comment (bjester) [reply kolibri/core/device/test/test_api.py]: There was indeed an issue which a test with this assertion would have illuminated.

**Review by bjester — APPROVED** [review]
Approved! :tada: Great work @AllanOXDi and co

I fixed a couple things that popped up during testing. I felt it would take longer to explain the issues than to throw in the fix, so let me know if you have any questions on my commit

**Pull request merged**
